### ISSUE TITLE
Support automatic fetching of ontologies from IRI upon CUDS deserialization

### ISSUE THREAD

**Issue by yoavnash** [simphony/osp-core#751]
A serialized CUDS object may be transferred across to network to a user that may not have the required ontologies installed on their local OSP-core. Therefore, an attempt to deserialize the received data will fail. However, if the ontologies are available online, then OSP-core can fetch the missing ontologies by attempting to download them automatically during the deserialization process.

Related to #660

**Comment by yoavnash** [comment]
@csadorf

**Comment by csadorf** [comment]
@yoavnash I think the actual issue that @CasperWa brought up, is that it is not possible to deserialize a CUDS container object in case that the ontologies cannot be downloaded at all, e.g., because the server is offline or there is no internet connection.

**Comment by CasperWA** [comment]
Indeed. In case one does not have the ontology/-ies locally and they are not available online (due to incorrect IRIs or servers being offline or other) then there is no way to serialize the JSON content to Python objects - I believe? It's essentially what you're writing in the OP as well - just with the addition of a complete halt in the serialization process if the information is not available anywhere.

At such a point, the only solution left is to manipulate the JSON object directly, which is far from ideal and defeats the purpose of this package (in my opinion).

**Comment by kysrpex** [comment]
I think _"automatic fetching of ontologies from IRI upon CUDS deserialization"_ is the wrong approach. Why would one bother downloading the resource from the internet, when the sender has the ontology installed? It makes more sense to send the resource (the ontology) over the network if the receiver does not have it. This is actually issue #438 in my opinion.

**Comment by CasperWA** [comment]
@kysrpex this is true. The issue here is also that deserialization is not at all possible if there is no ontology access whatsoever.

**Comment by yoavnash** [comment]
I updated the title and description. I hope it now better describes the main point that was raised. Thanks for the feedback.

**Comment by yoavnash** [comment]
> I think _"automatic fetching of ontologies from IRI upon CUDS deserialization"_ is the wrong approach. Why would one bother downloading the resource from the internet, when the sender has the ontology installed? It makes more sense to send the resource (the ontology) over the network if the receiver does not have it. This is actually issue #438 in my opinion.

Note that CUDS are not necessarily sent via the transport layer. Instead, they could be deserialized using the utility function and sent via other means.
